### PR TITLE
Connection ID Steering

### PR DESCRIPTION
**Pull request by victorstewart**
Under connection migration scenarios (ex. mobile client moving from LTE to WiFi) or intra-machine topology changes (ex. count of running server instances) in order to ensure stable packet steering there must be information in the connection ID to determine target machine + software server instance.

This patch allows the application to embed such information.

### CODE REVIEW

**Comment by victorstewart** [comment]
RE [https://github.com/litespeedtech/lsquic/issues/161#issue-703561967](https://github.com/litespeedtech/lsquic/issues/161#issue-703561967)

**Comment by dtikhonov** [comment]
Thanks, this looks good!  I will check out the patch later this week.

One thing I noticed is that the styling of the patch does not match the rest of the code (mostly whitespace).  Now, unlike some other open-source projects (* looks at just about *all* of them *), we won't make you go over your patch with a fine-toothed comb before deigning to accept it.  On the other hand, be warned that we will change the whitespace to suite our style -- that means that `git blame` won't find you (at least not on HEAD).  With that said, feel free to take action -- change the whitespace -- or leave it alone :slightly_smiling_face:

**Comment by victorstewart** [comment]
change any formatting you like! or cue me on the styling to follow and I'll make sure to follow suit in the future.

**Comment by dtikhonov** [comment]
The best style documentation is the code itself.

File: include/lsquic.h [hunk 1]
```diff
@@ -1186,6 +1186,12 @@ struct lsquic_engine_api
      */
     const struct lsquic_packout_mem_if  *ea_pmi;
     void                                *ea_pmi_ctx;
+
+    /**
+     * Optional interface to control the creation of connection IDs
+     */
+    void  (*ea_generate_cid)(lsquic_cid_t *);
```
Comment (dtikhonov) [inline include/lsquic.h:1193]: Was this a conscious decision on your part not to include `lsquic_conn_t *` in the list of arguments?  You may not need it in your use case, but having some way to get back at the context in the callback may be useful to others.

We should rename the function to `es_generate_scid` to indicate that these are *source* CIDs that are being generated.

File: src/liblsquic/lsquic_enc_sess_ietf.c [hunk 2]
```diff
@@ -553,8 +553,10 @@ gen_trans_params (struct enc_sess_iquic *enc_sess, unsigned char *buf,
             }
             cce->cce_seqno = seqno + 1;
             cce->cce_flags = CCE_SEQNO;
-            lsquic_generate_cid(&cce->cce_cid,
-                            enc_sess->esi_enpub->enp_settings.es_scid_len);
+
+            if (enc_sess->esi_enpub->enp_generate_cid) enc_sess->esi_enpub->enp_generate_cid(&cce->cce_cid);
```
Comment (dtikhonov) [inline src/liblsquic/lsquic_enc_sess_ietf.c:557]: I would pass `es_scid_len` to the callback as I originally proposed.  The reason for this is that `es_scid_len` is used in the packet parsing code: the generated SCIDs *must be* that long.

File: src/liblsquic/lsquic_engine.c [hunk 3]
```diff
@@ -576,6 +576,9 @@ lsquic_engine_new (unsigned flags,
     engine->pub.enp_lookup_cert  = api->ea_lookup_cert;
     engine->pub.enp_cert_lu_ctx  = api->ea_cert_lu_ctx;
     engine->pub.enp_get_ssl_ctx  = api->ea_get_ssl_ctx;
+
+    if (api->ea_generate_cid) engine->pub.enp_generate_cid = api->ea_generate_cid;
```
Comment (dtikhonov) [inline src/liblsquic/lsquic_engine.c:580]: Nit: this can be an unconditional assignment.

**Review by dtikhonov — CHANGES_REQUESTED** [review]

File: include/lsquic.h (hunk 1 above)
Comment (victorstewart) [reply include/lsquic.h:1193]: makes sense. done!

File: src/liblsquic/lsquic_enc_sess_ietf.c (hunk 2 above)
Comment (victorstewart) [reply src/liblsquic/lsquic_enc_sess_ietf.c:557]: done. i put it as `unsigned`. any reason we shouldn't use uint_fast8_t given the max connection ID length is 20?

File: src/liblsquic/lsquic_engine.c (hunk 3 above)
Comment (victorstewart) [reply src/liblsquic/lsquic_engine.c:580]: done

**Comment by victorstewart** [comment]
git add was **not** a fan of my whitespace usage either, but i appeased it

File: bin/test_common.c [hunk 4]
```diff
@@ -2043,7 +2043,7 @@ pba_init (struct packout_buf_allocator *pba, unsigned max)
 
 
 void *
-pba_allocate (void *packout_buf_allocator, void *peer_ctx, unsigned short size,
+pba_allocate (void *packout_buf_allocator, void *peer_ctx, lsquic_conn_ctx_t * conn_ctx, unsigned short size,
```
Comment (dtikhonov) [inline bin/test_common.c]: That seems to be from a different PR

Comment (victorstewart) [reply bin/test_common.c]: LOL i know. I accidentally forgot to change branches locally. also just when you don't think you have to compile locally, the github hall monitor slaps you across the face. i fixed all the errors on both PRs though.

File: src/liblsquic/lsquic_enc_sess_ietf.c (hunk 2 above)
Comment (dtikhonov) [reply src/liblsquic/lsquic_enc_sess_ietf.c]: It does not matter for the callback argument type.  We use integer types from `<stdint.h>` very sparingly in "lsquic.h".  Too, the type should match `es_scid_len`, which is `unsigned int`.

`lsquic_cid.len` is `uint_fast8_t`, which is what's important.

**Comment by dtikhonov** [comment]
> git add was not a fan of my whitespace usage either, but i appeased it

Eh, who cares what `git` thinks about whitespace?
